The core features of the package are shown in the following paragraphs.

In [1]:
import engicalc as ec
import numpy as np

## Integration of Pint units

### Common Units

Common units are stored as variables and are automatically imported with the packag The following units are available:

In [2]:
for unit in ec.units.items():
    print(unit[1])

kg
t
kNm
Nm
N
kN
MN
m
cm
dm
mm
km
rad
deg
%
‰
s
°C
K
MPa



### Handling Units 

Units can be added using operators:

In [3]:
v = 30 *ec.kNm

v_t = v + 30*ec.kNm


You can also access the full unit registry:

In [4]:
u = 400 *ec.ureg.hours

print(u)

400 h


Convert units using Pint:

In [5]:
u.to(ec.s)

<Quantity(1440000, 'second')>

## Parsing

### Cell Parsing

A simple parsing function extracts calculations from a cell:

In [6]:
v = 30 *ec.kNm

v_t = v + 30*ec.kNm

ec.parse_cell()

[{'variable_name': 'v',
  'expression': '30 * ec.kNm',
  'result': <Quantity(30, 'kNm')>},
 {'variable_name': 'v_t',
  'expression': 'v + 30 * ec.kNm',
  'result': <Quantity(60, 'kNm')>}]

#### Drawbacks

- The parsing function loses datatype information, returning rows as strings.
- Currently only assignments work, no conditionals, or other python syntax

In [7]:
b = 20 

if b >20:
    b = 40

ec.parse_cell()

[{'variable_name': 'b', 'expression': '20', 'result': 20},
 {'variable_name': 'b', 'expression': '40', 'result': 40}]

## Markdown rendering

Parsed cell content is processed using `sympy.sympify` and converted to `latex` via the `sympy` latexprinter. Eventually the `latex`code is inserted into a Markdown math environment. The `ec.render()`function is capable of the following:


### Numeric representation

The numeric representation shows the Variablename and its valu

In [8]:
v = 30 *ec.kNm

v_t = v + 30*ec.kNm


Theta_pl_A= 5 
m_u_A = 10 
m_y_A=5
l=1
b_w=0.1
EI_II=20
q_S1_A= 10

ec.render(symbolic=True)

$$\begin{aligned}v& = 30 \cdot \mathrm{kNm} = 30 \ \mathrm{kNm} \\ v_{t}& = v + 30 \cdot \mathrm{kNm} = 60 \ \mathrm{kNm} \\ \Theta_{pl A}& = 5 \\ m_{u A}& = 10 \\ m_{y A}& = 5 \\ l& = 1 \\ b_{w}& = 0.1 \\ EI_{II}& = 20 \\ q_{S1 A}& = 10\end{aligned}$$

### Symbolic Representation

The symbolic representation is also showing the calculation.

In [9]:
alpha_u_A = np.sqrt(Theta_pl_A) / 2
Delta = ((np.sin(alpha_u_A) + (m_u_A - m_y_A) * l * b_w / (3 * EI_II)) * 24 * EI_II / l**3)*ec.m
q_u_A = Delta + q_S1_A*ec.m
ec.render(raw=False)

$$\begin{aligned}\alpha_{u A}& = \frac{\sqrt{\Theta_{pl A}}}{2} = 1.12 \\ \Delta& = \left(\sin{\left(\alpha_{u A} \right)} + \frac{\left(m_{u A} - m_{y A}\right) \cdot l \cdot b_{w}}{3 \cdot EI_{II}}\right) \cdot 24 \cdot EI_{II} \cdot \frac{1}{l^{3}} \cdot \mathrm{m} = 435.64 \ \mathrm{m} \\ q_{u A}& = \Delta + q_{S1 A} \cdot \mathrm{m} = 445.64 \ \mathrm{m}\end{aligned}$$

#### Only symbolic

As a sidefunctionality, only the symbolic part can be shown.

In [10]:
Delta
l
x = 2*Delta
ec.render(symbolic=True, numeric=False)

expr
expr


$$\begin{aligned}\Delta& = \left(\sin{\left(\alpha_{u A} \right)} + \frac{\left(m_{u A} - m_{y A}\right) \cdot l \cdot b_{w}}{3 \cdot EI_{II}}\right) \cdot 24 \cdot EI_{II} \cdot \frac{1}{l^{3}} \cdot \mathrm{m} \\ l& = 1 \\ x& = 2 \cdot \Delta\end{aligned}$$

### Recalling variables

The defined variables in the notebook are stored in a container. They can be recalled at anytime and ec.rendered again.

In [11]:
v
Delta
ec.render()
ec.parse_cell()

expr
expr


$$\begin{aligned}v& = 30 \cdot \mathrm{kNm} = 30 \ \mathrm{kNm} \\ \Delta& = \left(\sin{\left(\alpha_{u A} \right)} + \frac{\left(m_{u A} - m_{y A}\right) \cdot l \cdot b_{w}}{3 \cdot EI_{II}}\right) \cdot 24 \cdot EI_{II} \cdot \frac{1}{l^{3}} \cdot \mathrm{m} = 435.64 \ \mathrm{m}\end{aligned}$$

expr
expr


[{'variable_name': 'v',
  'expression': '30 * ec.kNm',
  'result': <Quantity(30, 'kNm')>},
 {'variable_name': 'Delta',
  'expression': '(np.sin(alpha_u_A) + (m_u_A - m_y_A) * l * b_w / (3 * EI_II)) * 24 * EI_II / l ** 3 * ec.m',
  'result': <Quantity(435.63623, 'meter')>}]

In [12]:
from params import *

In [13]:
a
parse_cell()

expr


[]

### Multiple Rows

The markdown math environment is capable of displaying the equations in multiple rows:

In [14]:
v
Delta
Theta_pl_A
EI_II
v_t
ec.render(rows=3, symbolic=False)

ec.render(rows=5, symbolic=False)

expr
expr
expr
expr
expr


$$\begin{aligned}v& = 30 \ \mathrm{kNm} \quad & \Delta& = 435.64 \ \mathrm{m} \quad & \Theta_{pl A}& = 5 \\ EI_{II}& = 20 \quad & v_{t}& = 60 \ \mathrm{kNm} \quad & \end{aligned}$$

expr
expr
expr
expr
expr


$$\begin{aligned}v& = 30 \ \mathrm{kNm} \quad & \Delta& = 435.64 \ \mathrm{m} \quad & \Theta_{pl A}& = 5 \quad & EI_{II}& = 20 \quad & v_{t}& = 60 \ \mathrm{kNm}\end{aligned}$$

### Numpy functions

The package is based around the numpy functions and they should be used. Currently the `numpy`is stripped and the numpyfunction gets translated to `sympy` via `sympy.sympify`.

In [15]:
alpha = 45*ec.deg
test = np.atan(alpha.to(ec.los) + 25)

ec.render(raw=False)

$$\begin{aligned}\alpha& = 45 \cdot \mathrm{°} = 45 \ \mathrm{°} \\ test& = \operatorname{atan}{\left(\alpha + 25 \right)} = 1.53 \ \mathrm{rad}\end{aligned}$$

An array is translated to a matrix.

In [16]:
F_x = np.abs(np.array([-13,30,23,12])*ec.kN)
F_v = F_x * 2
F_z = F_x*np.atan(alpha)**np.sqrt(1)
F_y = F_x * F_v
ec.render(symbolic=True)

$$\begin{aligned}F_{x}& = \left|{\left[\begin{matrix}- 13 \cdot \mathrm{kN}\\30 \cdot \mathrm{kN}\\23 \cdot \mathrm{kN}\\12 \cdot \mathrm{kN}\end{matrix}\right]}\right| = \left[\begin{matrix}13\\30\\23\\12\end{matrix}\right] \ \mathrm{kN} \\ F_{v}& = F_{x} \cdot 2 = \left[\begin{matrix}26\\60\\46\\24\end{matrix}\right] \ \mathrm{kN} \\ F_{z}& = F_{x} \cdot \operatorname{atan}^{\sqrt{1}}{\left(\alpha \right)} = \left[\begin{matrix}8.66\\19.97\\15.31\\7.99\end{matrix}\right] \ \mathrm{kN} \cdot \mathrm{rad} \\ F_{y}& = F_{x} \cdot F_{v} = \left[\begin{matrix}338\\1800\\1058\\288\end{matrix}\right] \ \mathrm{kN}^{2}\end{aligned}$$

### Raw Markdown

As the markdowncode is stored anyways, it can be output aswell. Could be used to copy into a tabl

In [17]:
v
Delta
Theta_pl_A
ec.render(symbolic=True, raw=True)

expr
expr
expr
$$\begin{aligned}v& = 30 \cdot \mathrm{kNm} = 30 \ \mathrm{kNm} \\ \Delta& = \left(\sin{\left(\alpha_{u A} \right)} + \frac{\left(m_{u A} - m_{y A}\right) \cdot l \cdot b_{w}}{3 \cdot EI_{II}}\right) \cdot 24 \cdot EI_{II} \cdot \frac{1}{l^{3}} \cdot \mathrm{m} = 435.64 \ \mathrm{m} \\ \Theta_{pl A}& = 5\end{aligned}$$


$$\begin{aligned}v& = 30 \cdot \mathrm{kNm} = 30 \ \mathrm{kNm} \\ \Delta& = \left(\sin{\left(\alpha_{u A} \right)} + \frac{\left(m_{u A} - m_{y A}\right) \cdot l \cdot b_{w}}{3 \cdot EI_{II}}\right) \cdot 24 \cdot EI_{II} \cdot \frac{1}{l^{3}} \cdot \mathrm{m} = 435.64 \ \mathrm{m} \\ \Theta_{pl A}& = 5\end{aligned}$$


### Special Characters

Some special characters are inserted in the string before the sympy conversion takes plac In the `output` module, a replacement dictionary is created to replace the special characters. This can be expanded. It has to correspond to the `latex` syntax.

In [18]:
diam_infty = 20

infty__infty_infty__diam = 10
ec.render(raw=False)

$$\begin{aligned}\oslash_{\infty}& = 20 \\ \infty^{\infty \oslash}_{\infty}& = 10\end{aligned}$$

### Pint unit handling

In [19]:
v = v_t.to(ec.Nm) + 30*2*ec.Nm
q = v_t.magnitude*ec.Nm + v
q__shortcut = v_t.m * ec.Nm + v
ec.render()

$$\begin{aligned}v& = v_{t} + 30 \cdot 2 \cdot \mathrm{Nm} = 60060.0 \ \mathrm{Nm} \\ q& = v_{t} \cdot \mathrm{Nm} + v = 60120.0 \ \mathrm{Nm} \\ q^{shortcut}& = v_{t} \cdot \mathrm{Nm} + v = 60120.0 \ \mathrm{Nm}\end{aligned}$$

### Functions

It can be useful to display the calculations that have been done in a function environment. For that there is a second parsing function. The Funcion parses the local variables.

In [20]:
from IPython.display import Markdown

In [21]:
def test(alpha__top, b):
    x = alpha__top + b 
    y = alpha__top-b*2
    z = x + y
    display(Markdown('**Only Symbolic representation**'))
    ec.render_func(numeric=False ,rows=2)
    display(Markdown('**Whole representation**'))
    ec.render_func(numeric=True ,rows=3)
    return z



In [22]:
z = test(3,4)

**Only Symbolic representation**

$$\begin{aligned}\alpha^{top}& = 3 \quad & b& = 4 \\ x& = \alpha^{top} + b \quad & y& = \alpha^{top} - b \cdot 2 \\ z& = x + y \quad & \end{aligned}$$

**Whole representation**

$$\begin{aligned}\alpha^{top}& = 3 \quad & b& = 4 \quad & x& = \alpha^{top} + b = 7 \\ y& = \alpha^{top} - b \cdot 2 = -5 \quad & z& = x + y = 2 \quad & \end{aligned}$$

In [23]:
z__2 = test(5*ec.kNm + 3*ec.kNm, 3*ec.kNm)

**Only Symbolic representation**

$$\begin{aligned}\alpha^{top}& = 8 \ \mathrm{kNm} \quad & b& = 3 \ \mathrm{kNm} \\ x& = \alpha^{top} + b \quad & y& = \alpha^{top} - b \cdot 2 \\ z& = x + y \quad & \end{aligned}$$

**Whole representation**

$$\begin{aligned}\alpha^{top}& = 8 \ \mathrm{kNm} \quad & b& = 3 \ \mathrm{kNm} \quad & x& = \alpha^{top} + b = 11 \ \mathrm{kNm} \\ y& = \alpha^{top} - b \cdot 2 = 2 \ \mathrm{kNm} \quad & z& = x + y = 13 \ \mathrm{kNm} \quad & \end{aligned}$$

#### Drawback

The `ec.render` function parses the assignments inside the function. 

In [24]:
def tititi(a, b):
    x = a + b 
    y = a**b
    z = x + y
    return z

ec.render()

TypeError: loop of ufunc does not support argument 0 of type float which has no callable rint method

In [ ]:
z = tititi(3,4)

ec.render()

$$\begin{aligned}z& = \operatorname{tititi}{\left(3,4 \right)} = 88\end{aligned}$$

In [ ]:
L_K__y = 13.2 *ec.m

ec.render()

$$\begin{aligned}L^{y}_{K}& = 13.2 \cdot \mathrm{m} = 13.2 \ \mathrm{m}\end{aligned}$$

## Markdown tables

It can be useful to summarize the calculations in a tabl For that the variables can easily be inserted into the tabl Using the function `render_list`.

In [ ]:
import pandas as pd

# Example lists
col_1 = ec.render_list([z, v, q__shortcut, v_t], numeric=False, raw=True)
col_2 = ec.render_list([z, v, q__shortcut, v_t], symbolic=False, raw=True)
names = ['Höhe', 'Differenz', 'Test', 'Test3']

# Define column names
columnnames = ['Bezeichnung', 'Berechnung', 'Berechnung2']

# Create DataFrame
DF = pd.DataFrame(list(zip(names, col_1, col_2)), columns=columnnames)

# Display the DataFrame
display(Markdown(DF.to_markdown(tablefmt='pipe', index=False)))

| Bezeichnung   | Berechnung                                                                    | Berechnung2                                                           |
|:--------------|:------------------------------------------------------------------------------|:----------------------------------------------------------------------|
| Höhe          | $$\begin{aligned}z& = \operatorname{tititi}{\left(3,4 \right)}\end{aligned}$$ | $$\begin{aligned}z& = 88\end{aligned}$$                               |
| Differenz     | $$\begin{aligned}v& = v_{t} + 30 \cdot 2 \cdot \mathrm{Nm}\end{aligned}$$     | $$\begin{aligned}v& = 60060.0 \ \mathrm{Nm}\end{aligned}$$            |
| Test          | $$\begin{aligned}q^{shortcut}& = v_{t} \cdot \mathrm{Nm} + v\end{aligned}$$   | $$\begin{aligned}q^{shortcut}& = 60120.0 \ \mathrm{Nm}\end{aligned}$$ |
| Test3         | $$\begin{aligned}v_{t}& = v + 30 \cdot \mathrm{kNm}\end{aligned}$$            | $$\begin{aligned}v_{t}& = 60 \ \mathrm{kNm}\end{aligned}$$            |